integration/spark2/src/main/scala/org/apache/spark/sql/events/MergeBloomIndexEventListener.scala

```diff
@@ -24,59 +24,88 @@ import org.apache.spark.internal.Logging
+        val carbonTable = loadModel.getCarbonDataLoadSchema.getCarbonTable
+        val segmentId = loadModel.getSegmentId
+
+        // filter out bloom datamap, skip lazy datamap
+        val bloomDatamaps = DataMapStoreManager.getInstance().getAllDataMap(carbonTable).asScala
+          .filter(_.getDataMapSchema.getProviderName.equalsIgnoreCase(
+            DataMapClassProvider.BLOOMFILTER.getShortName))
+          .filter(!_.getDataMapSchema.isLazy).toList
+
+        mergeBloomIndex(sparkSession, carbonTable, bloomDatamaps, Seq(segmentId))
+
+      case compactPreStatusUpdateEvent: AlterTableCompactionPreStatusUpdateEvent =>
+        // For compact process, segment can not be accessed at this time
+        val carbonTable = compactPreStatusUpdateEvent.carbonTable
+        val mergedLoadName = compactPreStatusUpdateEvent.mergedLoadName
+        val segmentId = CarbonDataMergerUtil.getLoadNumberFromLoadName(mergedLoadName)
+
+        // filter out bloom datamap, skip lazy datamap
+        val bloomDatamaps = DataMapStoreManager.getInstance().getAllDataMap(carbonTable).asScala
+          .filter(_.getDataMapSchema.getProviderName.equalsIgnoreCase(
+            DataMapClassProvider.BLOOMFILTER.getShortName))
+          .filter(!_.getDataMapSchema.isLazy).toList
+
+        mergeBloomIndex(sparkSession, carbonTable, bloomDatamaps, Seq(segmentId))
+
       case datamapPostEvent: BuildDataMapPostExecutionEvent =>
-        LOGGER.info("Load post status event-listener called for merge bloom index")
+        // For rebuild datamap process, datamap is disabled when rebuilding
+        if (!datamapPostEvent.isFromRebuild || null == datamapPostEvent.dmName) {
+          // ignore datamapPostEvent from loading and compaction
+          return
+        }
+
         val carbonTableIdentifier = datamapPostEvent.identifier
         val carbonTable = DataMapStoreManager.getInstance().getCarbonTable(carbonTableIdentifier)
-        val tableDataMaps = DataMapStoreManager.getInstance().getAllDataMap(carbonTable)
-        val sparkSession = SparkSession.getActiveSession.get
 
-        // filter out bloom datamap
-        var bloomDatamaps = tableDataMaps.asScala.filter(
-          _.getDataMapSchema.getProviderName.equalsIgnoreCase(
+        // filter out current rebuilt bloom datamap
+        val bloomDatamaps = DataMapStoreManager.getInstance().getAllDataMap(carbonTable).asScala
+          .filter(_.getDataMapSchema.getProviderName.equalsIgnoreCase(
             DataMapClassProvider.BLOOMFILTER.getShortName))
-
-        if (datamapPostEvent.isFromRebuild) {
-          if (null != datamapPostEvent.dmName) {
-            // for rebuild process
-            bloomDatamaps = bloomDatamaps.filter(
-              _.getDataMapSchema.getDataMapName.equalsIgnoreCase(datamapPostEvent.dmName))
-          }
-        } else {
-          // for load process, skip lazy datamap
-          bloomDatamaps = bloomDatamaps.filter(!_.getDataMapSchema.isLazy)
```

Comment:
The reason for removing this branch is that bloom datamap is never lazy?  But line 51 and line 65 also handle the lazy scenes ？

Reply:
original implementation deal all three scenes(load/compact/rebuild), so use if-else to distinguish. Now, three scenes correspond to different event. Only rebuild deals with lazy bloom datamap